File: packages/efa_platform_client_ts/test/session.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import type { FetchLike } from "../src/lib/client";
import { AccountApiError, PlatformAuthRequiredError } from "../src/lib/errors";
import { PlatformSession } from "../src/lib/session";
import type { PlatformSessionStore } from "../src/lib/store";
import type { StoredPlatformSession } from "../src/lib/types";
import { fakeJwt } from "./helpers";

const ORIGIN = "https://api.example.test";

class MemoryStore implements PlatformSessionStore {
    value: StoredPlatformSession | null = null;
    read(): Promise<StoredPlatformSession | null> {
        return Promise.resolve(this.value);
    }
    write(session: StoredPlatformSession): Promise<void> {
        this.value = session;
        return Promise.resolve();
    }
    clear(): Promise<void> {
        this.value = null;
        return Promise.resolve();
    }
}

function storedSession(overrides: Partial<StoredPlatformSession> = {}): StoredPlatformSession {
    return {
        accessToken: fakeJwt("user-1"),
        refreshToken: "refresh-1",
        expiresAtMs: Date.now() + 600_000,
        email: "capsuleer@example.com",
        userId: "user-1",
        ...overrides,
    };
}

function tokenPairBody(userId = "user-1", refreshToken = "refresh-2") {
    return { accessToken: fakeJwt(userId), refreshToken, expiresIn: 900 };
}

function ok(body: unknown, status = 200) {
    return new Response(JSON.stringify(body), {
        status,
        headers: { "Content-Type": "application/json" },
    });
}

function authError(status: number, code: string) {
    return new Response(JSON.stringify({ error: code, message: code }), {
        status,
        headers: { "Content-Type": "application/json" },
    });
}

/** A fetch mock returning a fresh Response per call (Response bodies are single-use). */
function fetchReturning(body: () => Response) {
    return vi.fn().mockImplementation(() => Promise.resolve(body()));
}

function makeSession(
    store: MemoryStore,
    fetchFn: ReturnType<typeof vi.fn>,
    onAuthRequired?: () => void,
) {
    return new PlatformSession({
        origin: ORIGIN,
        store,
        fetchFn: fetchFn as unknown as FetchLike,
        onAuthRequired,
    });
}

afterEach(() => {
    vi.unstubAllGlobals();
});

describe("PlatformSession cold start", () => {
    it("reads as signed out with an empty store", async () => {
        const fetchFn = vi.fn();
        const session = makeSession(new MemoryStore(), fetchFn);
        await session.ready;
        expect(session.me).toBeNull();
        expect(fetchFn).not.toHaveBeenCalled();
    });

    it("rotates the stored session once and keeps the identity", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = fetchReturning(() => ok(tokenPairBody()));
        const session = makeSession(store, fetchFn);
        await session.ready;

        expect(fetchFn).toHaveBeenCalledTimes(1);
        expect(store.value?.refreshToken).toBe("refresh-2");
        expect(session.me).toEqual({ userId: "user-1", email: "capsuleer@example.com" });
    });

    it("clears a session whose refresh token the server rejects, without firing onAuthRequired", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = fetchReturning(() => authError(401, "invalid_token"));
        const onAuthRequired = vi.fn();
        const session = makeSession(store, fetchFn, onAuthRequired);
        await session.ready;

        expect(store.value).toBeNull();
        expect(session.me).toBeNull();
        expect(onAuthRequired).not.toHaveBeenCalled();
    });

    it("keeps the session when the cold-start refresh is offline", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = vi.fn().mockImplementation(() => Promise.reject(new TypeError("offline")));
        const session = makeSession(store, fetchFn);
        await session.ready;

        expect(store.value?.refreshToken).toBe("refresh-1");
        expect(session.me?.userId).toBe("user-1");
    });

    it("keeps the stored session when the rotated pair identifies another account", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = fetchReturning(() => ok(tokenPairBody("user-other")));
        const session = makeSession(store, fetchFn);
        await session.ready;

        expect(store.value?.refreshToken).toBe("refresh-1");
        expect(session.me?.userId).toBe("user-1");
    });
});

describe("PlatformSession auth flows", () => {
    it("login stores the pair and publishes the identity", async () => {
        const store = new MemoryStore();
        const fetchFn = fetchReturning(() => ok(tokenPairBody()));
        const session = makeSession(store, fetchFn);
        await session.ready;

        const identities: unknown[] = [];
        const unsubscribe = session.subscribeIdentity((v) => identities.push(v));
        await session.login("capsuleer@example.com", "password123");

        expect(store.value?.refreshToken).toBe("refresh-2");
        expect(session.me).toEqual({ userId: "user-1", email: "capsuleer@example.com" });
        expect(identities).toEqual([null, { userId: "user-1", email: "capsuleer@example.com" }]);
        unsubscribe();
    });

    it("rejects a login pair whose access token has no subject", async () => {
        const store = new MemoryStore();
        const fetchFn = fetchReturning(() =>
            ok({ accessToken: fakeJwt(null), refreshToken: "r", expiresIn: 900 }),
        );
        const session = makeSession(store, fetchFn);
        await session.ready;

        const error = await session.login("a@b.c", "password123").catch((e: unknown) => e);
        expect(error).toBeInstanceOf(AccountApiError);
        expect((error as AccountApiError).code).toBe("invalid_token");
        expect(store.value).toBeNull();
    });

    it("forwards the email locale on signup and password reset", async () => {
        const store = new MemoryStore();
        const fetchFn = fetchReturning(() => ok({ ok: true }));
        const session = new PlatformSession({
            origin: ORIGIN,
            store,
            fetchFn: fetchFn as unknown as FetchLike,
            emailLocale: () => "zh",
        });
        await session.ready;

        await session.signup("a@b.c", "password123");
        await session.requestPasswordReset("a@b.c");
        const bodies = fetchFn.mock.calls.map((c) =>
            JSON.parse((c as [string, RequestInit])[1].body as string),
        );
        expect(bodies).toEqual([
            { email: "a@b.c", password: "password123", locale: "zh" },
            { email: "a@b.c", locale: "zh" },
        ]);
    });

    it("logout revokes server-side best-effort and clears locally", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = vi
            .fn()
            .mockImplementationOnce(() => Promise.resolve(ok(tokenPairBody()))) // cold start
            .mockImplementationOnce(() => Promise.reject(new TypeError("offline"))); // logout fails
        const session = makeSession(store, fetchFn);
        await session.ready;

        await session.logout();

        expect(store.value).toBeNull();
        expect(session.me).toBeNull();
        expect(fetchFn.mock.calls.some((c) => String(c[0]).endsWith("/platform/auth/logout"))).toBe(
            true,
        );
    });

    it("deregister posts the password with a bearer token, then clears", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = fetchReturning(() => ok(tokenPairBody()));
        const session = makeSession(store, fetchFn);
        await session.ready;

        await session.deregister("password123");
        const deregisterCall = fetchFn.mock.calls.find((c) =>
            String(c[0]).endsWith("/platform/auth/deregister"),
        ) as [string, RequestInit];
        expect(deregisterCall).toBeDefined();
        expect((deregisterCall[1].headers as Record<string, string>).Authorization).toMatch(
            /^Bearer /,
        );
        expect(JSON.parse(deregisterCall[1].body as string)).toEqual({ password: "password123" });
        expect(store.value).toBeNull();
    });
});

describe("PlatformSession.authedFetch", () => {
    it("attaches the stored access token when it is still valid", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = fetchReturning(() => ok(tokenPairBody()));
        const session = makeSession(store, fetchFn);
        await session.ready;
        fetchFn.mockClear();

        fetchFn.mockImplementation(() => Promise.resolve(new Response("ok", { status: 200 })));
        const response = await session.authedFetch("https://api.example.test/platform/auth/x");
        expect(response.status).toBe(200);
        const headers = (fetchFn.mock.calls[0] as [string, RequestInit])[1].headers as Headers;
        expect(headers.get("Authorization")).toMatch(/^Bearer /);
    });

    it("refreshes an expired access token before the request", async () => {
        const store = new MemoryStore();
        store.value = storedSession({ expiresAtMs: Date.now() - 1000 });
        const fetchFn = vi
            .fn()
            // Cold start is offline, so the expired stored session stays in place.
            .mockImplementationOnce(() => Promise.reject(new TypeError("offline")))
            .mockImplementation(() => Promise.resolve(ok(tokenPairBody())));
        const session = makeSession(store, fetchFn);
        await session.ready;
        fetchFn.mockClear();

        await session.authedFetch("https://api.example.test/x");
        // One rotation plus the request itself, both through fetchFn.
        expect(fetchFn).toHaveBeenCalledTimes(2);
        expect(store.value?.refreshToken).toBe("refresh-2");
    });

    it("serializes concurrent rotations through the mutex", async () => {
        const store = new MemoryStore();
        store.value = storedSession({ expiresAtMs: Date.now() - 1000 });
        const fetchFn = vi
            .fn()
            .mockImplementationOnce(() => Promise.reject(new TypeError("offline")))
            .mockImplementation(() => Promise.resolve(ok(tokenPairBody())));
        const session = makeSession(store, fetchFn);
        await session.ready;
        fetchFn.mockClear();

        await Promise.all([
            session.authedFetch("https://api.example.test/a"),
            session.authedFetch("https://api.example.test/b"),
        ]);
        // The second caller reuses the pair rotated by the first: one
        // rotation plus the two requests, all through fetchFn.
        expect(fetchFn).toHaveBeenCalledTimes(3);
    });

    it("force-rotates once and retries when the server answers 401", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        const fetchFn = fetchReturning(() => ok(tokenPairBody()));
        const session = makeSession(store, fetchFn);
        await session.ready;
        fetchFn.mockClear();

        fetchFn
            .mockImplementationOnce(() => Promise.resolve(new Response("nope", { status: 401 })))
            .mockImplementationOnce(() => Promise.resolve(ok(tokenPairBody())))
            .mockImplementationOnce(() => Promise.resolve(new Response("ok", { status: 200 })));

        const response = await session.authedFetch("https://api.example.test/x");
        expect(response.status).toBe(200);
        // Rejected request, forced rotation, retried request.
        expect(fetchFn).toHaveBeenCalledTimes(3);
    });

    it("clears the session and fires onAuthRequired once when the rotation is rejected", async () => {
        const store = new MemoryStore();
        store.value = storedSession();
        // Cold-start refresh succeeds; later rotations are rejected.
        const fetchFn = vi
            .fn()
            .mockImplementationOnce(() => Promise.resolve(ok(tokenPairBody())))
            .mockImplementation(() => Promise.resolve(authError(401, "invalid_token")));
        const onAuthRequired = vi.fn();
        const session = makeSession(store, fetchFn, onAuthRequired);
        await session.ready;

        await expect(session.authedFetch("https://api.example.test/a")).rejects.toBeInstanceOf(
            PlatformAuthRequiredError,
        );
        await expect(session.authedFetch("https://api.example.test/b")).rejects.toBeInstanceOf(
            PlatformAuthRequiredError,
        );
        expect(store.value).toBeNull();
        expect(session.me).toBeNull();
        expect(onAuthRequired).toHaveBeenCalledTimes(1);
    });

    it("throws PlatformAuthRequiredError when signed out", async () => {
        const session = makeSession(new MemoryStore(), vi.fn());
        await session.ready;
        await expect(session.authedFetch("https://api.example.test/x")).rejects.toBeInstanceOf(
            PlatformAuthRequiredError,
        );
    });
});

File: packages/efa_platform_client/test/session_test.dart
@TestOn("vm")
library;

import "dart:convert";
import "dart:typed_data";

import "package:dio/dio.dart";
import "package:efa_platform_client/efa_platform_client.dart";
import "package:flutter_test/flutter_test.dart";

class _FakeAdapter implements HttpClientAdapter {
  _FakeAdapter(this._onFetch);

  final Future<ResponseBody> Function(RequestOptions options) _onFetch;

  @override
  Future<ResponseBody> fetch(
    RequestOptions options,
    Stream<Uint8List>? requestStream,
    Future<void>? cancelFuture,
  ) => _onFetch(options);

  @override
  void close({bool force = false}) {}
}

/// In-memory [PlatformSessionStore]; [writeError] simulates a secure-storage
/// failure that leaves the stored session untouched (atomic single write).
class _MemoryStore implements PlatformSessionStore {
  StoredPlatformSession? session;
  Exception? writeError;

  @override
  Future<StoredPlatformSession?> read() async => session;

  @override
  Future<void> write(StoredPlatformSession value) async {
    final error = writeError;
    if (error != null) throw error;
    session = value;
  }

  @override
  Future<void> clear() async => session = null;
}

String _jwt(String subject) {
  String segment(Object value) => base64Url.encode(utf8.encode(jsonEncode(value)));
  return "${segment({"alg": "HS256", "typ": "JWT"})}.${segment({"sub": subject, "tv": 0})}.sig";
}

Map<String, dynamic> _pair(String suffix, {String? subject}) => {
  "accessToken": _jwt(subject ?? "user-$suffix"),
  "refreshToken": "refresh-$suffix",
  "expiresIn": 900,
};

/// A rotation result for the seeded account: the access-token subject stays
/// the account's user id (`user-old`).
Map<String, dynamic> _rotatedPair(String refreshSuffix) =>
    _pair(refreshSuffix, subject: "user-old");

ResponseBody _json(Object body, [int status = 200, Map<String, List<String>>? headers]) =>
    ResponseBody.fromString(
      jsonEncode(body),
      status,
      headers: {
        Headers.contentTypeHeader: ["application/json"],
        ...?headers,
      },
    );

const _invalidToken = {"error": "invalid_token", "message": "missing or invalid access token"};

/// Scriptable transport for the auth endpoints plus arbitrary other paths
/// (see [authedHandler]).
class _Server {
  int refreshCalls = 0;
  Exception? refreshThrow;
  ResponseBody Function()? refreshResponse;
  ResponseBody Function()? loginResponse;
  String? loggedOutRefreshToken;
  String? deregisterBearer;
  String? deregisterPassword;

  /// Handles requests that are not auth-endpoint calls (the `authed` escape
  /// hatch, public reads).
  Future<ResponseBody> Function(RequestOptions options)? authedHandler;

  Future<ResponseBody> fetch(RequestOptions options) async {
    final path = options.path;
    if (path.endsWith("/platform/auth/login")) {
      return loginResponse!();
    }
    if (path.endsWith("/platform/auth/refresh")) {
      refreshCalls++;
      final error = refreshThrow;
      if (error != null) throw error;
      return refreshResponse!();
    }
    if (path.endsWith("/platform/auth/logout")) {
      loggedOutRefreshToken = (options.data as Map<String, dynamic>)["refreshToken"] as String;
      return _json({"ok": true});
    }
    if (path.endsWith("/platform/auth/deregister")) {
      deregisterBearer = options.headers["Authorization"] as String?;
      deregisterPassword = (options.data as Map<String, dynamic>)["password"] as String?;
      return _json({"ok": true});
    }
    final handler = authedHandler;
    if (handler != null) return handler(options);
    throw StateError("unexpected request: $path");
  }
}

void main() {
  const origin = "https://test.invalid";
  late _MemoryStore store;
  late _Server server;
  int authRequiredCalls = 0;

  PlatformSession session({
    bool trackAuthRequired = true,
    String? cfAccessClientId,
    String? cfAccessClientSecret,
  }) => PlatformSession(
    origin: origin,
    store: store,
    dioFactory: () => Dio(BaseOptions())..httpClientAdapter = _FakeAdapter(server.fetch),
    cfAccessClientId: cfAccessClientId,
    cfAccessClientSecret: cfAccessClientSecret,
    onAuthRequired: trackAuthRequired ? () => authRequiredCalls++ : null,
  );

  StoredPlatformSession storedSession({bool expired = false}) => StoredPlatformSession(
    accessToken: _jwt("user-old"),
    refreshToken: "refresh-old",
    expiresAt: expired
        ? DateTime.now().subtract(const Duration(minutes: 5))
        : DateTime.now().add(const Duration(minutes: 10)),
    email: "capsuleer@example.com",
    userId: "user-old",
  );

  void seedSignedIn({bool expired = false}) => store.session = storedSession(expired: expired);

  setUp(() {
    store = _MemoryStore();
    server = _Server()..refreshResponse = () => _json(_rotatedPair("boot"));
    authRequiredCalls = 0;
  });

  group("store contract", () {
    test("write/read/clear round-trips through the in-memory store", () async {
      final record = storedSession();
      await store.write(record);
      expect((await store.read())?.refreshToken, "refresh-old");
      await store.clear();
      expect(await store.read(), isNull);
    });
  });

  group("cold start", () {
    test("starts signed out without a stored session", () async {
      final s = session();
      await s.ready;

      expect(s.me, isNull);
      expect(server.refreshCalls, 0);
      expect(await s.identity.first, isNull);
    });

    test("startup refresh rotates the stored pair once per cold start", () async {
      seedSignedIn();
      final s = session();
      await s.ready;

      expect(s.me, const PlatformIdentity(userId: "user-old", email: "capsuleer@example.com"));
      expect(server.refreshCalls, 1);
      expect(store.session?.refreshToken, "refresh-boot");
    });

    test("startup refresh keeps the session when the server is unreachable", () async {
      seedSignedIn();
      server.refreshThrow = Exception("offline");
      final s = session();
      await s.ready;

      expect(s.me, isNotNull);
      expect(store.session?.refreshToken, "refresh-old");
    });

    test("startup refresh keeps the stored session when persisting the rotation fails", () async {
      seedSignedIn();
      store.writeError = Exception("secure storage unavailable");
      final s = session();
      await s.ready;

      // The rotation succeeded server-side but could not be persisted: the
      // stored pair stays intact (no partial session state) and the session
      // keeps working on it.
      expect(s.me, isNotNull);
      expect(server.refreshCalls, 1);
      expect(store.session?.refreshToken, "refresh-old");
    });

    test("startup refresh signs out when the refresh token is dead", () async {
      seedSignedIn();
      server.refreshResponse = () => _json(_invalidToken, 401);
      final s = session();
      await s.ready;

      expect(s.me, isNull);
      expect(store.session, isNull);
      // No login redirect at cold start: nothing is in flight that it would
      // recover.
      expect(authRequiredCalls, 0);
    });

    test("startup refresh keeps the stored session when the rotated pair has no subject", () async {
      seedSignedIn();
      server.refreshResponse = () =>
          _json({"accessToken": "not-a-jwt", "refreshToken": "refresh-bad", "expiresIn": 900});
      final s = session();
      await s.ready;

      // The malformed pair is never persisted: the stored session stays
      // intact and the state keeps identifying the prior account.
      expect(s.me?.userId, "user-old");
      expect(server.refreshCalls, 1);
      expect(store.session?.refreshToken, "refresh-old");
    });

    test(
      "startup refresh keeps the stored session when the rotated pair subject mismatches",
      () async {
        seedSignedIn();
        server.refreshResponse = () => _json(_pair("other"));
        final s = session();
        await s.ready;

        // A pair identifying another account must not replace the stored
        // access token while the state still identifies the prior account.
        expect(s.me?.userId, "user-old");
        expect(server.refreshCalls, 1);
        expect(store.session?.refreshToken, "refresh-old");
      },
    );
  });

  group("auth flows", () {
    test("login stores the session and publishes the identity", () async {
      server.loginResponse = () => _json(_pair("1"));
      final s = session();
      await s.ready;
      final identities = <PlatformIdentity?>[];
      final sub = s.identity.listen(identities.add);

      await s.login(email: "capsuleer@example.com", password: "secret-pw");
      // Identity events are delivered asynchronously.
      await Future<void>.delayed(Duration.zero);

      expect(s.me, const PlatformIdentity(userId: "user-1", email: "capsuleer@example.com"));
      expect(store.session?.refreshToken, "refresh-1");
      expect(identities, [
        isNull,
        const PlatformIdentity(userId: "user-1", email: "capsuleer@example.com"),
      ]);
      await sub.cancel();
    });

    test("a login failure leaves the session signed out", () async {
      server.loginResponse = () =>
          _json({"error": "invalid_credentials", "message": "invalid email or password"}, 401);
      final s = session();
      await s.ready;

      await expectLater(
        () => s.login(email: "a@b.c", password: "wrong"),
        throwsA(isA<AccountApiException>()),
      );
      expect(s.me, isNull);
      expect(store.session, isNull);
    });

    test("login rejects a token pair without a usable JWT subject", () async {
      seedSignedIn();
      server.loginResponse = () =>
          _json({"accessToken": "not-a-jwt", "refreshToken": "refresh-bad", "expiresIn": 900});
      final s = session();
      await s.ready;

      await expectLater(
        () => s.login(email: "capsuleer@example.com", password: "secret-pw"),
        throwsA(isA<AccountApiException>()),
      );

      // The malformed pair is never stored and the identity is cleared.
      expect(store.session, isNull);
      expect(s.me, isNull);
    });

    test("logout revokes the stored refresh token and clears the identity", () async {
      server.loginResponse = () => _json(_pair("1"));
      final s = session();
      await s.ready;
      await s.login(email: "capsuleer@example.com", password: "secret-pw");

      await s.logout();

      expect(server.loggedOutRefreshToken, "refresh-1");
      expect(s.me, isNull);
      expect(store.session, isNull);
    });

    test("deregister reuses a valid access token without refreshing", () async {
      server.loginResponse = () => _json(_pair("1"));
      final s = session();
      await s.ready;
      await s.login(email: "capsuleer@example.com", password: "secret-pw");

      await s.deregister(password: "secret-pw");

      expect(server.refreshCalls, 0);
      expect(server.deregisterBearer, "Bearer ${_jwt("user-1")}");
      expect(server.deregisterPassword, "secret-pw");
      expect(s.me, isNull);
      expect(store.session, isNull);
    });
  });

  group("expiry refresh", () {
    /// Drains the startup refresh, then re-seeds an expired pair so the
    /// refresh counter only measures the operation under test.
    Future<PlatformSession> sessionWithExpiredPair() async {
      seedSignedIn(expired: true);
      final s = session();
      await s.ready;
      server.refreshCalls = 0;
      store.session = storedSession(expired: true);
      return s;
    }

    test("deregister refreshes an expired access token and stores the rotated pair", () async {
      final s = await sessionWithExpiredPair();
      server.refreshResponse = () => _json(_rotatedPair("new"));

      await s.deregister(password: "secret-pw");

      // The operation rotates exactly once; a duplicate refresh would rotate
      // the refresh token twice and could invalidate the stored pair. A
      // successful deregistration clears the rotated pair again.
      expect(server.refreshCalls, 1);
      expect(server.deregisterBearer, "Bearer ${_jwt("user-old")}");
      expect(store.session, isNull);
    });

    test("concurrent operations are serialized and rotate the pair only once", () async {
      final s = await sessionWithExpiredPair();
      server.refreshResponse = () => _json(_rotatedPair("new"));

      // Both calls enter the refresh path while the stored pair is expired.
      // The second one must observe the rotated pair on its re-read inside
      // the critical section instead of refreshing again with the same (now
      // server-side dead) refresh token.
      await Future.wait([s.deregister(password: "secret-pw"), s.deregister(password: "secret-pw")]);

      expect(server.refreshCalls, 1);
      expect(store.session, isNull);
    });

    test("a refresh rejected as invalid clears the session and fires onAuthRequired", () async {
      final s = await sessionWithExpiredPair();
      server.refreshResponse = () => _json(_invalidToken, 401);

      await expectLater(
        () => s.deregister(password: "secret-pw"),
        throwsA(isA<PlatformAuthRequiredException>()),
      );
      expect(s.me, isNull);
      expect(store.session, isNull);
      expect(authRequiredCalls, 1);
    });

    test("expiry refresh rejects a rotated pair without a usable JWT subject", () async {
      final s = await sessionWithExpiredPair();
      server.refreshResponse = () =>
          _json({"accessToken": "not-a-jwt", "refreshToken": "refresh-bad", "expiresIn": 900});

      await expectLater(
        () => s.deregister(password: "secret-pw"),
        throwsA(isA<PlatformAuthRequiredException>()),
      );

      // The malformed pair is neither stored nor returned, and the doomed
      // session (its refresh token was rotated server-side) is cleared.
      expect(server.refreshCalls, 1);
      expect(server.deregisterBearer, isNull);
      expect(store.session, isNull);
    });

    test("expiry refresh rejects a rotated pair with a mismatched JWT subject", () async {
      final s = await sessionWithExpiredPair();
      server.refreshResponse = () => _json(_pair("other"));

      await expectLater(
        () => s.deregister(password: "secret-pw"),
        throwsA(isA<PlatformAuthRequiredException>()),
      );

      // A pair identifying another account is neither stored nor returned,
      // and the doomed session is cleared like a rejected refresh.
      expect(server.refreshCalls, 1);
      expect(server.deregisterBearer, isNull);
      expect(store.session, isNull);
    });
  });

  group("createComment", () {
    test("posts the body with a valid access token and decodes the comment", () async {
      seedSignedIn();
      RequestOptions? captured;
      server.authedHandler = (options) async {
        captured = options;
        return _json({
          "commentId": "c-1",
          "postId": "p-1",
          "authorId": "user-old",
          "authorDeleted": false,
          "body": "**hello**",
          "createdAt": "2026-08-28T00:00:00.000Z",
        }, 201);
      };
      final s = session();
      await s.ready;

      final comment = await s.createComment(postId: "p-1", body: "**hello**");

      expect(captured?.path, "$origin/platform/internal/posts/p-1/comments");
      expect(captured?.method, "POST");
      expect(captured?.headers["Authorization"], "Bearer ${_jwt("user-old")}");
      expect((captured?.data as Map<String, dynamic>)["body"], "**hello**");
      expect(comment.commentId, "c-1");
      expect(comment.authorId, "user-old");
      expect(comment.authorDeleted, isFalse);
      expect(comment.body, "**hello**");
    });

    test("maps a 403 to a PlatformApiException with the envelope code", () async {
      seedSignedIn();
      server.authedHandler = (options) async =>
          _json({"error": "forbidden", "message": "missing permission: comment:create"}, 403);
      final s = session();
      await s.ready;

      await expectLater(
        () => s.createComment(postId: "p-1", body: "hi"),
        throwsA(
          isA<PlatformApiException>()
              .having((e) => e.statusCode, "statusCode", 403)
              .having((e) => e.code, "code", "forbidden"),
        ),
      );
    });

    test("without a session throws PlatformAuthRequiredException", () async {
      final s = session();
      await s.ready;

      await expectLater(
        () => s.createComment(postId: "p-1", body: "hi"),
        throwsA(isA<PlatformAuthRequiredException>()),
      );
      expect(authRequiredCalls, 1);
    });
  });

  group("authed", () {
    const authedPath = "$origin/platform/internal/authed";

    test("attaches a valid access token without refreshing", () async {
      seedSignedIn();
      String? authorization;
      server.authedHandler = (options) async {
        authorization = options.headers["Authorization"] as String?;
        return _json({"ok": true});
      };
      final s = session();
      await s.ready;

      final data = await s.authed(
        (dio) async => (await dio.get<Map<String, dynamic>>(authedPath)).data,
      );

      expect(data, {"ok": true});
      expect(authorization, "Bearer ${_jwt("user-old")}");
      // Only the cold-start rotation ran.
      expect(server.refreshCalls, 1);
    });

    test("refreshes an expired access token before the request", () async {
      seedSignedIn(expired: true);
      String? authorization;
      server.authedHandler = (options) async {
        authorization = options.headers["Authorization"] as String?;
        return _json({"ok": true});
      };
      final s = session();
      await s.ready;
      server.refreshCalls = 0;
      store.session = storedSession(expired: true);
      server.refreshResponse = () => _json(_rotatedPair("new"));

      await s.authed((dio) => dio.get<Map<String, dynamic>>(authedPath));

      expect(server.refreshCalls, 1);
      expect(authorization, "Bearer ${_jwt("user-old")}");
      expect(store.session?.refreshToken, "refresh-new");
    });

    test("a 401 forces one rotation and retries the request once", () async {
      seedSignedIn();
      final seen = <String?>[];
      server
        ..authedHandler = (options) async {
          seen.add(options.headers["Authorization"] as String?);
          // Reject the pre-rotation token, accept the rotated one.
          return seen.length == 1 ? _json(_invalidToken, 401) : _json({"ok": true});
        }
        ..refreshResponse = () => _json(_rotatedPair("new"));
      final s = session();
      await s.ready;
      // Discount the cold-start rotation.
      server.refreshCalls = 0;

      final data = await s.authed(
        (dio) async => (await dio.get<Map<String, dynamic>>(authedPath)).data,
      );

      expect(data, {"ok": true});
      expect(server.refreshCalls, 1);
      expect(seen, ["Bearer ${_jwt("user-old")}", "Bearer ${_jwt("user-old")}"]);
      expect(store.session?.refreshToken, "refresh-new");
    });

    test("a 401 with a rejected refresh clears the session and throws", () async {
      seedSignedIn();
      server
        ..authedHandler = ((options) async => _json(_invalidToken, 401))
        ..refreshResponse = () => _json(_invalidToken, 401);
      final s = session();
      await s.ready;

      await expectLater(
        () => s.authed((dio) => dio.get<Map<String, dynamic>>(authedPath)),
        throwsA(isA<PlatformAuthRequiredException>()),
      );
      expect(store.session, isNull);
      expect(s.me, isNull);
      expect(authRequiredCalls, 1);
    });

    test("without a session the request never leaves the client", () async {
      var requests = 0;
      server.authedHandler = (options) async {
        requests++;
        return _json({"ok": true});
      };
      final s = session();
      await s.ready;

      await expectLater(
        () => s.authed((dio) => dio.get<Map<String, dynamic>>(authedPath)),
        throwsA(isA<PlatformAuthRequiredException>()),
      );
      expect(requests, 0);
      expect(authRequiredCalls, 1);
    });

    test("onAuthRequired is throttled per signed-out stretch and reset by login", () async {
      seedSignedIn();
      server
        ..authedHandler = ((options) async => _json(_invalidToken, 401))
        ..refreshResponse = () => _json(_invalidToken, 401);
      final s = session();
      await s.ready;

      // A burst of failing requests triggers one navigation, not many.
      for (var i = 0; i < 3; i++) {
        await expectLater(
          () => s.authed((dio) => dio.get<Map<String, dynamic>>(authedPath)),
          throwsA(isA<PlatformAuthRequiredException>()),
        );
      }
      expect(authRequiredCalls, 1);

      // The next successful login rearms the hook.
      server.loginResponse = () => _json(_pair("1"));
      await s.login(email: "capsuleer@example.com", password: "secret-pw");
      server.refreshResponse = () => _json(_invalidToken, 401);
      await expectLater(
        () => s.authed((dio) => dio.get<Map<String, dynamic>>(authedPath)),
        throwsA(isA<PlatformAuthRequiredException>()),
      );
      expect(authRequiredCalls, 2);
    });
  });

  group("public reads", () {
    test("never attach credentials or trigger auth logic", () async {
      seedSignedIn();
      String? authorization;
      server.authedHandler = (options) async {
        authorization = options.headers["Authorization"] as String?;
        return _json({
          "postId": "p",
          "authorId": null,
          "authorDeleted": true,
          "fitHash": "abc",
          "createdAt": "2026-08-19T00:00:00.000Z",
          "commentCount": 0,
        });
      };
      final s = session();
      await s.ready;

      final record = await s.getPost("p");

      expect(record?.fitHash, "abc");
      expect(authorization, isNull);
      expect(authRequiredCalls, 0);
    });
  });

  group("Cloudflare Access service token", () {
    test("rides on authed requests and public reads, not just auth calls", () async {
      seedSignedIn();
      final seen = <String, RequestOptions>{};
      server.authedHandler = (options) async {
        seen[options.path] = options;
        if (options.path.endsWith("/posts/p")) {
          return _json({
            "postId": "p",
            "authorId": null,
            "authorDeleted": true,
            "fitHash": "abc",
            "createdAt": "2026-08-19T00:00:00.000Z",
            "commentCount": 0,
          });
        }
        return _json({"ok": true});
      };
      final s = session(cfAccessClientId: "cf-id-1.access", cfAccessClientSecret: "cf-secret-1");
      await s.ready;

      await s.authed(
        (dio) async => (await dio.get<Map<String, dynamic>>("$origin/platform/internal/x")).data,
      );
      await s.getPost("p");

      for (final options in seen.values) {
        expect(options.headers["CF-Access-Client-Id"], "cf-id-1.access");
        expect(options.headers["CF-Access-Client-Secret"], "cf-secret-1");
      }
      expect(seen, hasLength(2));
    });

    test("is absent from every client when not configured", () async {
      seedSignedIn();
      final seen = <RequestOptions>[];
      server.authedHandler = (options) async {
        seen.add(options);
        return _json({"ok": true});
      };
      final s = session();
      await s.ready;

      await s.authed(
        (dio) async => (await dio.get<Map<String, dynamic>>("$origin/platform/internal/x")).data,
      );

      expect(seen.single.headers["CF-Access-Client-Id"], isNull);
      expect(seen.single.headers["CF-Access-Client-Secret"], isNull);
    });
  });
}
